ENH Adds infrequent categories support to OrdinalEncoder (#25677)
scikit-learn/scikit-learn: doc/modules/preprocessing.rst

See :ref:`dict_feature_extraction` for categorical features that are
represented as a dict, not as scalars.

.. _encoder_infrequent_categories:

Infrequent categories
---------------------

> :class:`OneHotEncoder` and :class:`OrdinalEncoder` supports aggregating

Reviewer: ```suggestion
:class:`OneHotEncoder` and :class:`OrdinalEncoder` support aggregating
```

   input feature. `max_categories` includes the feature that combines
   infrequent categories.

In the following example with :class:`OrdinalEncoder`, the categories `'dog' and
'snake'` are considered infrequent::

   >>> X = np.array([['dog'] * 5 + ['cat'] * 20 + ['rabbit'] * 10 +
   ...               ['snake'] * 3], dtype=object).T
   >>> enc = preprocessing.OrdinalEncoder(min_frequency=6).fit(X)
   >>> enc.infrequent_categories_
   [array(['dog', 'snake'], dtype=object)]
   >>> enc.transform(np.array([['dog'], ['cat'], ['rabbit'], ['snake']]))
   array([[2.],
          [0.],
          [1.],
          [2.]])

> :class:`OrdinalEncoder` can be configured with an encoding unknown categories and

Reviewer: ```suggestion
:class:`OrdinalEncoder` can be configured to encode unknown categories and
```
Not quite sure if this is what you wanted to say, but the original seems to be missing words or have too many?

Author: Yea, I agree the original wording was unclear. I reworked the whole paragraph.

   input feature. `max_categories` includes the feature that combines
   infrequent categories.

In the following example with :class:`OrdinalEncoder`, the categories `'dog' and
'snake'` are considered infrequent::

   >>> X = np.array([['dog'] * 5 + ['cat'] * 20 + ['rabbit'] * 10 +
   ...               ['snake'] * 3], dtype=object).T
   >>> enc = preprocessing.OrdinalEncoder(min_frequency=6).fit(X)
   >>> enc.infrequent_categories_
   [array(['dog', 'snake'], dtype=object)]
   >>> enc.transform(np.array([['dog'], ['cat'], ['rabbit'], ['snake']]))
   array([[2.],
          [0.],
          [1.],
          [2.]])

:class:`OrdinalEncoder`'s `max_categories` do **not** take into account missing
or unknown categories. Setting `unknown_value` or `encoded_missing_value` to an
integer will increase the number of unique integer codes by one each. This can
result in up to `max_categories + 2` integer codes. In the following example,
"a" and "d" are considered infrequent and grouped together into a single

> category, "b" and "c" are their own categories, unknown values are encoded as 3

Reviewer 2: Can you explain the motivation of treating unknown values different from infrequent ones? Or is a simplification for now?

Author: It's a simplification to get infrequent categories without adding more API. To treat unknown categories as the infrequent ones, we'll need to add a new `handle_unknown="infrequent_if_exist"`.

Reviewer 2: ok that's fair :)

and missing values are encoded as 4.

  >>> X_train = np.array(
  ...     [["a"] * 5 + ["b"] * 20 + ["c"] * 10 + ["d"] * 3 + [np.nan]],
  ...     dtype=object).T
  >>> enc = preprocessing.OrdinalEncoder(
  ...     handle_unknown="use_encoded_value", unknown_value=3,
  ...     max_categories=3, encoded_missing_value=4)
  >>> _ = enc.fit(X_train)
  >>> X_test = np.array([["a"], ["b"], ["c"], ["d"], ["e"], [np.nan]], dtype=object)
  >>> enc.transform(X_test)
  array([[2.],
         [0.],
         [1.],
         [2.],
         [3.],
         [4.]])

> Similarity, :class:`OneHotEncoder` can be configured the same way to group together

Reviewer 2: But here unknown values are either grouped with the infrequent ones or "ignored" which is not really possible in OrdinalEncoder.

Author: I see why this is confusing coming from the previous paragraph. I removed "the same way", because the next part explains what happens with unknown categories in `OneHotEncoder`.

   input feature. `max_categories` includes the feature that combines
   infrequent categories.

In the following example with :class:`OrdinalEncoder`, the categories `'dog' and
'snake'` are considered infrequent::

   >>> X = np.array([['dog'] * 5 + ['cat'] * 20 + ['rabbit'] * 10 +
   ...               ['snake'] * 3], dtype=object).T
   >>> enc = preprocessing.OrdinalEncoder(min_frequency=6).fit(X)
   >>> enc.infrequent_categories_
   [array(['dog', 'snake'], dtype=object)]
   >>> enc.transform(np.array([['dog'], ['cat'], ['rabbit'], ['snake']]))
   array([[2.],
          [0.],
          [1.],
          [2.]])

:class:`OrdinalEncoder`'s `max_categories` do **not** take into account missing
or unknown categories. Setting `unknown_value` or `encoded_missing_value` to an
integer will increase the number of unique integer codes by one each. This can
result in up to `max_categories + 2` integer codes. Categories are encoded by

> frequency and therefore less frequent categories care grouped together. In the

Reviewer 3: "Categories are encoded by frequency" is not true in the current state of this PR. The default category mapping is still the lexicographical (alphabetical + digits) order of the category labels.

The frequency based ordering was a suggestion for a follow-up PR. The purpose of this PR is only to collapsed the least frequent categories into their own category.
